tools/checkpoint_conversion/convert_roberta_checkpoints.py

```diff
@@ -0,0 +1,352 @@
+    "roberta_large": ("roberta.large", "roberta-large"),
+}
+
+DOWNLOAD_SCRIPT_URL = "https://dl.fbaipublicfiles.com/fairseq/models/{}.tar.gz"
+
+EXTRACT_DIR = "./{}"
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string(
+    "preset", None, f'Must be one of {",".join(PRESET_MAP.keys())}'
+)
+
+
+def download_model(preset, size, hf_model_name):
+    print("-> Download original weights.")
+    extract_dir = EXTRACT_DIR.format(size)
+    archive_file_path = keras.utils.get_file(
+        fname=None,
+        origin=DOWNLOAD_SCRIPT_URL.format(size),
+        cache_subdir=os.path.join("checkpoint_conversion", preset),
+    )
+
+    os.system(f"tar -xvf {archive_file_path}")
+
+    # The original `tar.gz` file does not have the vocab files. Let's fetch
+    # them from HF.
+    vocabulary_path = keras.utils.get_file(
+        fname=None,
+        origin=f"https://huggingface.co/{hf_model_name}/raw/main/vocab.json",
+    )
+    shutil.copy(vocabulary_path, extract_dir)
+    merges_path = keras.utils.get_file(
+        fname=None,
+        origin=f"https://huggingface.co/{hf_model_name}/raw/main/merges.txt",
+    )
+    shutil.copy(merges_path, extract_dir)
+
+
+def convert_checkpoints(preset, size):
+    print("\n-> Convert original weights to KerasNLP format.")
+    # RoBERTa paths.
+    extract_dir = EXTRACT_DIR.format(size)
+    checkpoint_path = os.path.join(extract_dir, "model.pt")
+
+    # Load PyTorch RoBERTa checkpoint.
+    pt_ckpt = torch.load(checkpoint_path, map_location=torch.device("cpu"))
+    pt_cfg = pt_ckpt["args"]
+    pt_model = pt_ckpt["model"]
+
+    cfg = {
+        "num_layers": pt_cfg.encoder_layers,
+        "num_heads": pt_cfg.encoder_attention_heads,
+        "hidden_dim": pt_cfg.encoder_embed_dim,
+        "intermediate_dim": pt_cfg.encoder_ffn_embed_dim,
+        "dropout": pt_cfg.dropout,
+        "max_sequence_length": pt_cfg.max_positions,
+        "vocab_size": (
+            pt_model["decoder.sentence_encoder.embed_tokens.weight"]
+            .numpy()
+            .shape[0]
+        ),
+    }
+    print("Config:", cfg)
+
+    keras_nlp_model = keras_nlp.models.Roberta.from_preset(
+        preset, load_weights=False
+    )
+
+    # Embedding Layer.
+    keras_nlp_model.get_layer("embeddings").token_embedding.embeddings.assign(
+        pt_model["decoder.sentence_encoder.embed_tokens.weight"].numpy()
+    )
+    keras_nlp_model.get_layer(
+        "embeddings"
+    ).position_embedding.position_embeddings.assign(
+        pt_model["decoder.sentence_encoder.embed_positions.weight"].numpy()[
+            2:, :
+        ]
+    )
+
+    # Embedding LayerNorm.
+    keras_nlp_model.get_layer("embeddings_layer_norm").gamma.assign(
+        pt_model["decoder.sentence_encoder.emb_layer_norm.weight"].numpy()
+    )
+    keras_nlp_model.get_layer("embeddings_layer_norm").beta.assign(
+        pt_model["decoder.sentence_encoder.emb_layer_norm.bias"].numpy()
+    )
+
+    range_1 = (0, cfg["hidden_dim"])
```

Comment:
hard to understand these range variables, could you add a comment explaining what they are doing?

Reply:
Perfect! Thanks!